src/mscorlib/src/System/StubHelpers.cs

```diff
@@ -125,6 +125,112 @@ static internal void ClearNative(IntPtr pNative)
         }
     }  // class CSTRMarshaler
 
+    [ReliabilityContract(Consistency.WillNotCorruptState, Cer.MayFail)]
+    internal static class UTF8Marshaler
+    {
+        const int UTF8_CHAR_SIZE = 3;
+        static internal unsafe IntPtr ConvertToNative(int flags, string strManaged, IntPtr pNativeBuffer)
+        {
+            if (null == strManaged)
+            {
+                return IntPtr.Zero;
+            }
+            StubHelpers.CheckStringLength(strManaged.Length);
+
+            int nb;
+            byte* pbNativeBuffer = (byte*)pNativeBuffer;
+
+            // If we are marshaling into a stack buffer allocated by the ILStub
+            // we will use a "1-pass" mode where we convert the string directly into the unmanaged buffer.   
+            // else we will allocate the precise native heap memory.          
+            if (pbNativeBuffer != null)
+            {
+                // this is the number of bytes allocated by the ILStub.
+                nb = (strManaged.Length + 1) * UTF8_CHAR_SIZE;
+
+                // +1 for the '\0' that we add
+                nb += 1;
+
+                // nb is the actual number of bytes written by Encoding.GetBytes.
+                // use nb to de-limit the string since we are allocating more than 
+                // required on stack
+                nb = strManaged.GetBytesFromEncoding(pbNativeBuffer, nb, Encoding.UTF8);
+            }
+            // required bytes > 260 , allocate required bytes on heap             
+            else
+            {
+                nb = Encoding.UTF8.GetByteCount(strManaged);
```

Comment:
There is some interesting trade-off here. We could allocate more than enough memory (len+1)*MAX_UTF_8_SIZE, and avoid the extra pass. We do this in our ANSI marshaling, where SystemMaxDBCSCharSize could be anywhere between 1 and 3, and it's not clear to me if there is a clear choice, and we need to make a choice regardless. My first thinking is that we should optimize this to be fast, since majority of the string operation are transient ([in] only). If there are concerns with memory/working set, developers should be able to change marshaling signature to IntPtr and use our Marshal helpers. We can also build in some magic number to fallback to slower (but more memory efficient path), but it's not clear to me if that's actually better than not having that magic number at all.